avaframe/com1DFA/DFAfunctionsCython.pyx

```diff
@@ -1118,59 +1181,68 @@ def computeGradC(cfg, particles, headerNeighbourGrid, headerNormalGrid, double[:
-  cdef double dx, dy, dz, dn, r, hr, dwdr, wKernel
+  cdef double dx, dy, dz, dux, duy, duz, dn, r, hr, dwdr, wKernel
   cdef int Lx0, Ly0, iCell
   cdef double w[4]
   cdef int lInd, rInd
   cdef int indx, indy
-  cdef int j, ic, n, p, l, imax, imin, iPstart, iPend
+  cdef int k, ic, n, p, l, imax, imin, iPstart, iPend
   cdef int grad = gradient
+  # artificial viscosity parameters
+  cdef double dwdrr, area
+  cdef double pikl, flux
+  cdef double epsilon = 100
+  cdef double hk, hl, ck, cl, lambdakl
 
   # loop on particles
-  for j in range(N):
+  for k in range(N):
     gradhX = 0
     gradhY = 0
     gradhZ = 0
+    pikl = 0
     G1 = 0
     G2 = 0
     m11 = 0
     m12 = 0
     m22 = 0
-    x = xArray[j]
-    y = yArray[j]
-    z = zArray[j]
-    ux = uxArray[j]
-    uy = uyArray[j]
-    uz = uzArray[j]
+    x = xArray[k]
+    y = yArray[k]
+    z = zArray[k]
+    ux = uxArray[k]
+    uy = uyArray[k]
+    uz = uzArray[k]
+    hk = hArray[k]
+    mk = mass[k]
+
     # locate particle in SPH grid
     indx = <int>math.round(x / cszNeighbourGrid)
     indy = <int>math.round(y / cszNeighbourGrid)
 
     if SPHoption > 1:
-      # get normal vector
-      Lx0, Ly0, iCell, w[0], w[1], w[2], w[3] = getCellAndWeights(x, y, nColsNormal, nRowsNormal, cszNormal, interpOption)
-      nx, ny, nz = getVector(Lx0, Ly0, w[0], w[1], w[2], w[3], nxArray, nyArray, nzArray)
-      nx, ny, nz = normalize(nx, ny, nz)
-      # projection of gravity on normal vector
-      gravAcc3 = scalProd(nx, ny, nz, 0, 0, gravAcc)
-      uMag = norm(ux, uy, uz)
-      if uMag < velMagMin:
-          ux = 1
-          uy = 0
-          uz = -(1*nx + 0*ny) / nz
-          ux, uy, uz = normalize(ux, uy, uz)
-          K1 = 1
-          K2 = 1
-      else:
-          ux, uy, uz = normalize(ux, uy, uz)
-
-      uxOrtho, uyOrtho, uzOrtho = crossProd(nx, ny, nz, ux, uy, uz)
-      uxOrtho, uyOrtho, uzOrtho = normalize(uxOrtho, uyOrtho, uzOrtho)
-
-      g1 = nx/(nz)
-      g2 = ny/(nz)
+        # get normal vector
+        Lx0, Ly0, iCell, w[0], w[1], w[2], w[3] = getCellAndWeights(x, y, nColsNormal, nRowsNormal, cszNormal, interpOption)
+        nx, ny, nz = getVector(Lx0, Ly0, w[0], w[1], w[2], w[3], nxArray, nyArray, nzArray)
+        nx, ny, nz = normalize(nx, ny, nz)
+        # projection of gravity on normal vector
+        gravAcc3 = scalProd(nx, ny, nz, 0, 0, gravAcc)
+        if SPHoption > 2:
+            uMag = norm(ux, uy, uz)
+            if uMag < velMagMin:
+                ux = 1
+                uy = 0
+                uz = -(1*nx + 0*ny) / nz
+                ux, uy, uz = normalize(ux, uy, uz)
+                K1 = 1
+                K2 = 1
+            else:
+                ux, uy, uz = normalize(ux, uy, uz)
+
+                uxOrtho, uyOrtho, uzOrtho = crossProd(nx, ny, nz, ux, uy, uz)
+                uxOrtho, uyOrtho, uzOrtho = normalize(uxOrtho, uyOrtho, uzOrtho)
+
+                g1 = nx/(nz)
+                g2 = ny/(nz)
```

Comment:
This part is not covered by a test, but is quite a substantial part, please add a test